delegation/delegated.gno

```diff
@@ -0,0 +1,254 @@
+package delegation
+
+import (
+	"std"
+
+	"gno.land/p/demo/grc/grc20"
+	u256 "gno.land/p/demo/uint256"
+)
+
+// IDelegatedToken defines the interface for a delegated token.
+type IDelegatedToken interface {
+	GetStaker() std.Address
+	GetDelegatedTo(owner std.Address) std.Address
+	Delegate(owner, to std.Address)
+	Deposit()
+	DepositAmount(amount *u256.Uint)
+	Withdraw()
+	WithdrawAmount(amount *u256.Uint)
+}
+
+// DelegatedToken represents a token that can be delegated to another address.
+type DelegatedToken struct {
+	// staker is the address of the entity staking the tokens.
+	staker      std.Address
+
+	// token is the unerlying GRC20 token being staked.
+	token       grc20.Token
+
+	// delegatedTo maps owner address to their delegate address.
+	delegatedTo map[std.Address]std.Address
+
+	// balances keeps track of the delegated token balances for each account.
+	balances    map[std.Address]*u256.Uint
+
+	// allowances tracks the spending allowances for delegated tokens.
+	allowances  map[std.Address]map[std.Address]*u256.Uint
+
+	// totoalSupply represents the total amount of delegated tokens in circulation.
+	totalSupply *u256.Uint
+
+	// name is the name of the delegated token.
+	name        string
+
+	// symbol is the symbol of the delegated token.
+	symbol      string
+}
+
+var gDelegatedToken *DelegatedToken
+
+// Init initializes the DelegatedToken contract
+func Init(staker std.Address, token grc20.Token, name string, symbol string) {
+	gDelegatedToken = &DelegatedToken{
+		staker:      staker,
+		token:       token,
+		delegatedTo: make(map[std.Address]std.Address),
+		balances:    make(map[std.Address]*u256.Uint),
+		totalSupply: u256.Zero(),
+		name:        name,
+		symbol:      symbol,
+	}
+}
+
+// GetStaker returns the addrress of the staker that this delegated token wrapper uses.
+func GetStaker() std.Address {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+	return gDelegatedToken.staker
+}
+
+// GetDelegatedTo returns the address to whom the owner is delegated.
+func GetDelegatedTo(owner std.Address) std.Address {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	delegatee, exists := gDelegatedToken.delegatedTo[owner]
+	if !exists {
+		// if not explicitly delegated, return owner
+		return owner
+	}
+
+	return delegatee
+}
+
+func Delegate(owner, to std.Address) {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	currDelegatee := GetDelegatedTo(owner)
+
+	// if the new delegation is different from the current one
+	if currDelegatee != to {
+		gDelegatedToken.delegatedTo[owner] = to
+
+		// TODO: move delegates ?
```

Comment:
move to where? for what??

Reply:
I used to write it for notes, so it's not that important. removed